PeerDAS implementation

dapplion opened the pull request:
This is an umbrella PR for a PeerDAS implementation.

The plan is to merge smaller PRs from this one into unstable, bit by bit. The PeerDAS feature will be gated behind a config parameter and can coexist with stable code.

**Tracking issue** ➡️  
- https://github.com/sigp/lighthouse/issues/4983

jimmygchen committed:
1D PeerDAS prototype: Data format and Distribution (#5050)

* Build and publish column sidecars. Add stubs for gossip.

* Add blob column subnets

* Add `BlobColumnSubnetId` and initial compute subnet logic.

* Subscribe to blob column subnets.

* Introduce `BLOB_COLUMN_SUBNET_COUNT` based on DAS configuration parameter changes.

* Fix column sidecar type to use `VariableList` for data.

* Fix lint errors.

* Update types and naming to latest consensus-spec #3574.

* Fix test and some cleanups.

jimmygchen committed:
Merge branch 'unstable' into das

jimmygchen committed:
Merge branch 'unstable' into das

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	consensus/types/src/chain_spec.rs

jimmygchen committed:
Add `DataColumnSidecarsByRoot ` req/resp protocol (#5196)

* Add stub for `DataColumnsByRoot`

* Add basic implementation of serving RPC data column from DA checker.

* Store data columns in early attester cache and blobs db.

* Apply suggestions from code review

Co-authored-by: Eitan Seri-Levi <eserilev@gmail.com>
Co-authored-by: Jacob Kaufmann <jacobkaufmann18@gmail.com>

* Fix build.

* Store `DataColumnInfo` in database and various cleanups.

* Update `DataColumnSidecar` ssz max size and remove panic code.

---------

Co-authored-by: Eitan Seri-Levi <eserilev@gmail.com>
Co-authored-by: Jacob Kaufmann <jacobkaufmann18@gmail.com>

jacobkaufmann committed:
feat: add DAS KZG in data col construction (#5210)

* feat: add DAS KZG in data col construction

* refactor data col sidecar construction

* refactor: add data cols to GossipVerifiedBlockContents

* Disable windows tests for `das` branch. (c-kzg doesn't build on windows)

* Formatting and lint changes only.

* refactor: remove iters in construction of data cols

* Update vec capacity and error handling.

* Add `data_column_sidecar_computation_seconds` metric.

---------

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	.github/workflows/test-suite.yml
#	beacon_node/lighthouse_network/src/types/topics.rs

jacobkaufmann committed:
fix: update data col subnet count from 64 to 32 (#5413)

jacobkaufmann committed:
feat: add peerdas custody field to ENR (#5409)

* feat: add peerdas custody field to ENR

* add hash prefix step in subnet computation

* refactor test and fix possible u64 overflow

* default to min custody value if not present in ENR

jimmygchen committed:
Merge branch 'unstable' into das

jimmygchen committed:
Merge branch 'unstable' into das-unstable-merge-0415

# Conflicts:
#	Cargo.lock
#	beacon_node/beacon_chain/src/data_availability_checker.rs
#	beacon_node/beacon_chain/src/data_availability_checker/availability_view.rs
#	beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs
#	beacon_node/beacon_chain/src/data_availability_checker/processing_cache.rs
#	beacon_node/lighthouse_network/src/rpc/methods.rs
#	beacon_node/network/src/network_beacon_processor/mod.rs
#	beacon_node/network/src/sync/block_lookups/tests.rs
#	crypto/kzg/Cargo.toml

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into das

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into das

jimmygchen committed:
Fix merge conflicts.

jimmygchen committed:
Send custody data column to `DataAvailabilityChecker` for determining block importability (#5570)

* Only import custody data columns after publishing a block.

* Add `subscribe-all-data-column-subnets` and pass custody column count to `availability_cache`.

* Add custody requirement checks to `availability_cache`.

* Fix config not being passed to DAChecker and add more logging.

* Introduce `peer_das_epoch` and make blobs and columns mutually exclusive.

* Add DA filter for PeerDAS.

* Fix data availability check and use test_logger in tests.

* Fix subscribe to all data column subnets not working correctly.

* Fix tests.

* Only publish column sidecars if PeerDAS is activated. Add `PEER_DAS_EPOCH` chain spec serialization.

* Remove unused data column index in `OverflowKey`.

* Fix column sidecars incorrectly produced when there are no blobs.

* Re-instate index to `OverflowKey::DataColumn` and downgrade noisy debug log to `trace`.

dapplion committed:
DAS sampling on sync (#5616)

* Data availability sampling on sync

* Address @jimmygchen review

* Trigger sampling

* Address some review comments and only send `SamplingBlock` sync message after PEER_DAS_EPOCH.

---------

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	Cargo.lock
#	Cargo.toml
#	beacon_node/beacon_chain/src/block_verification.rs
#	beacon_node/http_api/src/publish_blocks.rs
#	beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
#	beacon_node/lighthouse_network/src/rpc/protocol.rs
#	beacon_node/lighthouse_network/src/types/pubsub.rs
#	beacon_node/network/src/sync/block_lookups/single_block_lookup.rs
#	beacon_node/store/src/hot_cold_store.rs
#	consensus/types/src/beacon_state.rs
#	consensus/types/src/chain_spec.rs
#	consensus/types/src/eth_spec.rs

jimmygchen committed:
Merge branch 'unstable' into das

dapplion committed:
Re-process early sampling requests (#5569)

* Re-process early sampling requests

# Conflicts:
#	beacon_node/beacon_processor/src/work_reprocessing_queue.rs
#	beacon_node/lighthouse_network/src/rpc/methods.rs
#	beacon_node/network/src/network_beacon_processor/rpc_methods.rs

* Update beacon_node/beacon_processor/src/work_reprocessing_queue.rs

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

* Add missing var

* Beta compiler fixes and small typo fixes.

* Remove duplicate method.

---------

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

dapplion committed:
Merge remote-tracking branch 'sigp/unstable' into das

jimmygchen committed:
Fix merge conflict.

jimmygchen committed:
Add data columns by root to currently supported protocol list (#5678)

* Add data columns by root to currently supported protocol list.

* Add missing data column by roots handling.

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	Cargo.lock
#	Cargo.toml
#	beacon_node/network/src/sync/block_lookups/tests.rs
#	beacon_node/network/src/sync/manager.rs

jimmygchen committed:
Fix simulator tests on `das` branch (#5731)

* Bump genesis delay in sim tests as KZG setup takes longer for DAS.

* Fix incorrect YAML spacing.

eserilev committed:
DataColumnByRange boilerplate (#5353)

* add boilerplate

* fmt

dapplion committed:
PeerDAS custody lookup sync (#5684)

* Implement custody sync

* Lint

* Fix tests

* Fix rebase issue

jimmygchen committed:
Add data column kzg verification and update `c-kzg`. (#5701)

* Add data column kzg verification and update `c-kzg`.

* Fix incorrect `Cell` size.

* Add kzg verification on rpc blocks.

* Add kzg verification on rpc data columns.

jimmygchen commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -2935,6 +3076,33 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+        }
+
+        // TODO(das): custody column SSE event
+        // TODO(das): Why is the slot necessary here?
```
I guess it's mainly used for debugging purpose
`MissingComponents(slot, root)` is mainly used for logging. I think it wouldn't hurt to add the slot here, 0 might be confusing.

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -145,6 +184,27 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
+        };
+
+        // TODO(das): report which column is invalid for proper peer scoring
+        let verified_custody_columns = custody_columns
```
we could do batch KZG verification here

jimmygchen replied:
ah, we can't really with the current construct, as they are done in the individual `KzgVerifiedDataColumn` constructors. We'd probably need an extra struct to hold the list if we need want this.

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs:
```diff
@@ -238,6 +331,17 @@ impl<E: EthSpec> PendingComponents<E> {
             block,
             blobs: Some(verified_blobs),
             blobs_available_timestamp,
+            // TODO(das) Do we need a check here for number of expected custody columns?
```
This check has already been done in `is_available` so this is redundant here. I'll remove the TODO

jimmygchen committed:
Rename `PEER_DAS_EPOCH` to `EIP7594_FORK_EPOCH` for client interop. (#5750)

dapplion committed:
Fetch custody columns in range sync (#5747)

* Fetch custody columns in range sync

* Clean up todos

jimmygchen committed:
Remove `BlobSidecar` construction and publish after PeerDAS activated (#5759)

* Avoid building and publishing blob sidecars after PeerDAS.

* Ignore gossip blobs with a slot greater than peer das activation epoch.

* Only attempt to verify blob count and import blobs before PeerDAS.

jimmygchen committed:
#5684 review comments (#5748)

* #5684 review comments.

* Doc and message update only.

* Fix incorrect condition when constructing `RpcBlock` with `DataColumn`s

dapplion committed:
Make sampling tests deterministic (#5775)

jimmygchen committed:
PeerDAS spec tests (#5772)

* Add get_custody_columns spec tests.

* Add kzg merkle proof spec tests.

* Add SSZ spec tests.

* Add remaining KZG tests

* Load KZG only once per process, exclude electra tests and add missing SSZ tests.

* Fix lint and missing changes.

* Ignore macOS generated file.

dapplion committed:
Merge remote branch 'sigp/unstable' into das

dapplion committed:
Merge remote tracking branch 'origin/unstable' into das

dapplion committed:
Implement unconditional reconstruction for supernodes (#5781)

* Implement unconditional reconstruction for supernodes

* Move code into KzgVerifiedCustodyDataColumn

* Remove expect

* Add test

* Thanks justin

dapplion committed:
Add withhold attack mode for interop (#5788)

* Add withhold attack mode

* Update readme

* Drop added readmes

* Undo styling changes

jimmygchen committed:
Add column gossip verification and handle unknown parent block (#5783)

* Add column gossip verification and handle missing parent for columns.

* Review PR

* Fix rebase issue

* more lint issues :)

---------

Co-authored-by: dapplion <35266934+dapplion@users.noreply.github.com>

dapplion committed:
Trigger sampling on sync events (#5776)

* Trigger sampling on sync events

* Update beacon_chain.rs

* Fix tests

* Fix tests

jimmygchen committed:
PeerDAS parameter changes for devnet-0 (#5779)

* Update PeerDAS parameters to latest values.

* Lint fix

* Fix lint.

dapplion committed:
Update hardcoded subnet count to 64 (#5791)

jimmygchen committed:
Fix incorrect columns per subnet and config cleanup (#5792)

* Tidy up PeerDAS preset and config values.

* Fix broken config

jimmygchen committed:
Fix DAS branch CI (#5793)

* Fix invalid syntax.

* Update cli doc. Ignore get_custody_columns test temporarily.

* Fix failing test and add verify inclusion test.

* Undo accidentally removed code.

jimmygchen committed:
Only attempt reconstruct columns once. (#5794)

dapplion committed:
Re-enable precompute table for peerdas kzg (#5795)

dapplion committed:
Merge branch 'unstable' into das

jimmygchen committed:
Update subscription filter. (#5797)

jimmygchen committed:
Remove penalty for duplicate columns (expected due to reconstruction) (#5798)

jimmygchen committed:
Revert DAS config for interop testing. Optimise get_custody_columns function. (#5799)

jimmygchen committed:
Don't perform reconstruction for proposer node as it already has all the columns. (#5806)

dapplion committed:
Multithread compute_cells_and_proofs (#5805)

* Multi-thread reconstruct data columns

* Multi-thread path for block production

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	.github/workflows/test-suite.yml
#	beacon_node/network/src/sync/block_lookups/mod.rs
#	beacon_node/network/src/sync/block_lookups/single_block_lookup.rs
#	beacon_node/network/src/sync/network_context.rs

jimmygchen committed:
Fix CI errors.

jimmygchen committed:
Move PeerDAS type-level config to configurable `ChainSpec` (#5828)

* Move PeerDAS type level config to `ChainSpec`.

* Fix tests

dapplion committed:
Misc custody lookup improvements (#5821)

* Improve custody requests

* Type DataColumnsByRootRequestId

* Prioritize peers and load balance

* Update tests

* Address PR review

jimmygchen committed:
Merge branch 'unstable' into das

jimmygchen committed:
Rename deploy_block in network config (`das` branch) (#5852)

* Rename deploy_block.txt to deposit_contract_block.txt

* fmt

---------

Co-authored-by: Pawan Dhananjay <pawandhananjay@gmail.com>

dapplion committed:
Merge branch 'unstable' into das

jimmygchen committed:
Fix CI and merge issues.

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs
#	lcli/src/main.rs

dapplion committed:
Store data columns individually in store and caches (#5890)

* Store data columns individually in store and caches

* Implement data column pruning

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	Cargo.lock

jimmygchen committed:
Update reconstruction benches to newer criterion version. (#5949)

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	.github/workflows/test-suite.yml

kevaundray committed:
chore: add `recover_cells_and_compute_proofs` method (#5938)

* chore: add recover_cells_and_compute_proofs method

* Introduce type alias `CellsAndKzgProofs` to address type complexity.

---------

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

jimmygchen committed:
Update `csc` format in ENR and spec tests for devnet-1 (#5966)

* Update `csc` format in ENR.

* Add spec tests for `recover_cells_and_kzg_proofs`.

* Add tests for ENR.

* Fix failing tests.

* Add protection against invalid csc value in ENR.

* Fix lint

jimmygchen committed:
Fix csc encoding and decoding (#5997)

jimmygchen committed:
Fix data column rpc request not being sent due to incorrect limits set. (#6000)

jimmygchen committed:
Fix incorrect inbound request count causing rate limiting. (#6025)

jimmygchen committed:
Merge branch 'stable' into das

# Conflicts:
#	beacon_node/network/src/sync/block_lookups/tests.rs
#	beacon_node/network/src/sync/block_sidecar_coupling.rs
#	beacon_node/network/src/sync/manager.rs
#	beacon_node/network/src/sync/network_context.rs
#	beacon_node/network/src/sync/network_context/requests.rs

dapplion committed:
Merge remote-tracking branch 'unstable' into das

jimmygchen committed:
Add kurtosis config for DAS testing (#5968)

* Add kurtosis config for DAS testing.

* Fix invalid yaml file

* Update network parameter files.

kevaundray committed:
chore: add rust PeerdasKZG crypto library for peerdas functionality and rollback c-kzg dependency to 4844 version (#5941)

* chore: add recover_cells_and_compute_proofs method

* chore: add rust peerdas crypto library

* chore: integrate peerdaskzg rust library into kzg crate

* chore(multi):

- update `ssz_cell_to_crypto_cell`
- update conversion from the crypto cell type to a Vec<u8>. Since the Rust library defines them as references to an array, the conversion is simply `to_vec`

* chore(multi):

- update rest of code to handle the new crypto `Cell` type
- update test case code to no longer use the Box type

* chore: cleanup of superfluous conversions

* chore: revert c-kzg dependency back to v1

* chore: move dependency into correct order

* chore: update rust dependency

- This version includes a new method `PeerDasContext::with_num_threads`

* chore: remove Default initialization of PeerDasContext and explicitly set the parameters in `new_from_trusted_setup`

* chore: cleanup exports

* chore: commit updated cargo.lock

* Update Cargo.toml

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

* chore: rename dependency

* chore: update peerdas lib

- sets the blst version to 0.3 so that it matches whatever lighthouse is using. Although 0.3.12 is latest, lighthouse is pinned to 0.3.3

* chore: fix clippy lifetime

- Rust doesn't allow you to elide the lifetime on type aliases

* chore: cargo clippy fix

* chore: cargo fmt

* chore: update lib to add redundant checks (these will be removed in consensus-specs PR 3819)

* chore: update dependency to ignore proofs

* chore: update peerdas lib to latest

* update lib

* chore: remove empty proof parameter

---------

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

jimmygchen committed:
Update PeerDAS interop testnet config (#6069)

* Update interop testnet config.

* Fix typo and remove target peers

jimmygchen committed:
Avoid retrying same sampling peer that previously failed. (#6084)

jimmygchen committed:
Various fixes to custody range sync  (#6004)

* Only start requesting batches when there are good peers across all custody columns to avoid spaming block requests.

* Add custody peer check before mutating `BatchInfo` to avoid inconsistent state.

* Add check to cover a case where batch is not processed while waiting for custody peers to become available.

* Fix lint and logic error

* Fix `good_peers_on_subnet` always returning false for `DataColumnSubnet`.

* Add test for `get_custody_peers_for_column`

* Revert epoch parameter refactor.

* Fall back to default custody requiremnt if peer ENR is not present.

* Add metrics and update code comment.

* Add more debug logs.

* Use subscribed peers on subnet before MetaDataV3 is implemented. Remove peer_id matching when injecting error because multiple peers are used for range requests. Use randomized custodial peer to avoid repeatedly sending requests to failing peers. Batch by range request where possible.

* Remove unused code and update docs.

* Add comment

kevaundray committed:
chore: update peerdas-kzg library (#6118)

* chore: update peerDAS lib

* chore: update library

* chore: update library to version that include "init context" benchmarks and optional validation checks

* chore: (can remove) -- Add benchmarks for init context

AgeManning committed:
Prevent continuous searchers for low-peer networks (#6162)

dapplion committed:
Merge branch 'unstable' into das

dapplion committed:
Fix merge conflicts

jimmygchen committed:
Add cli flag to enable sampling and disable by default. (#6209)

kevaundray committed:
chore: Use reference to an array representing a blob instead of an owned KzgBlob (#6179)

* add KzgBlobRef type

* modify code to use KzgBlobRef

* clippy

* Remove Deneb blob related changes to maintain compatibility with `c-kzg-4844`.

---------

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

jimmygchen committed:
Store computed custody subnets in PeerDB and fix custody lookup test (#6218)

* Fix failing custody lookup tests.

* Store custody subnets in PeerDB, fix custody lookup test and refactor some methods.

jimmygchen committed:
Merge branch 'unstable' into das

# Conflicts:
#	beacon_node/beacon_chain/src/beacon_chain.rs
#	beacon_node/beacon_chain/src/block_verification_types.rs
#	beacon_node/beacon_chain/src/builder.rs
#	beacon_node/beacon_chain/src/data_availability_checker.rs
#	beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs
#	beacon_node/beacon_chain/src/data_column_verification.rs
#	beacon_node/beacon_chain/src/early_attester_cache.rs
#	beacon_node/beacon_chain/src/historical_blocks.rs
#	beacon_node/beacon_chain/tests/store_tests.rs
#	beacon_node/lighthouse_network/src/discovery/enr.rs
#	beacon_node/network/src/service.rs
#	beacon_node/src/cli.rs
#	beacon_node/store/src/hot_cold_store.rs
#	beacon_node/store/src/lib.rs
#	lcli/src/generate_bootnode_enr.rs

jimmygchen committed:
Fix CI failures after merge.

ackintosh committed:
Batch sampling requests by peer (#6256)

* Batch sampling requests by peer

* Fix clippy errors

* Fix tests

* Add column_index to error message for ease of tracing

* Remove outdated comment

jimmygchen committed:
Fix range sync never evaluating request as finished, causing it to get stuck. (#6276)

jimmygchen committed:
Merge branch 'unstable' into das-0821-merge

# Conflicts:
#	Cargo.lock
#	Cargo.toml
#	beacon_node/beacon_chain/src/beacon_chain.rs
#	beacon_node/beacon_chain/src/data_availability_checker.rs
#	beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs
#	beacon_node/beacon_chain/src/data_column_verification.rs
#	beacon_node/beacon_chain/src/kzg_utils.rs
#	beacon_node/beacon_chain/src/metrics.rs
#	beacon_node/beacon_processor/src/lib.rs
#	beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs
#	beacon_node/lighthouse_network/src/rpc/config.rs
#	beacon_node/lighthouse_network/src/rpc/methods.rs
#	beacon_node/lighthouse_network/src/rpc/outbound.rs
#	beacon_node/lighthouse_network/src/rpc/rate_limiter.rs
#	beacon_node/lighthouse_network/src/service/api_types.rs
#	beacon_node/lighthouse_network/src/types/globals.rs
#	beacon_node/network/src/network_beacon_processor/mod.rs
#	beacon_node/network/src/network_beacon_processor/rpc_methods.rs
#	beacon_node/network/src/network_beacon_processor/sync_methods.rs
#	beacon_node/network/src/sync/block_lookups/common.rs
#	beacon_node/network/src/sync/block_lookups/mod.rs
#	beacon_node/network/src/sync/block_lookups/single_block_lookup.rs
#	beacon_node/network/src/sync/block_lookups/tests.rs
#	beacon_node/network/src/sync/manager.rs
#	beacon_node/network/src/sync/network_context.rs
#	consensus/types/src/data_column_sidecar.rs
#	crypto/kzg/Cargo.toml
#	crypto/kzg/benches/benchmark.rs
#	crypto/kzg/src/lib.rs

jimmygchen committed:
Fix custody tests and load PeerDAS KZG instead.

jimmygchen committed:
Fix ef tests and bench compilation.

jimmygchen committed:
Fix failing sampling test.

jimmygchen committed:
Merge pull request #6287 from jimmygchen/das-0821-merge

Merge `unstable` into `das` 20240821

jimmygchen commented on .github/workflows/test-suite.yml:
```diff
@@ -88,37 +88,38 @@ jobs:
+#  release-tests-windows:
+#    name: release-tests-windows
+#    needs: [check-labels]
+#    if: needs.check-labels.outputs.skip_ci != 'true'
```
We should be able to uncomment this now as we use the prod `c-kzg-4844` version on the `das` branch.

jimmygchen replied:
Done

jimmygchen commented on testing/simulator/src/fallback_sim.rs:
```diff
@@ -15,7 +15,7 @@ use tokio::time::sleep;
 
 const END_EPOCH: u64 = 16;
-const GENESIS_DELAY: u64 = 32;
+const GENESIS_DELAY: u64 = 60;
```
We might be able to revert this now we use prod `c-kzg-4844` and `rust-eth-kzg` for PeerDAS, which only takes 500ms to initialize.

jimmygchen replied:
Done

jimmygchen commented on testing/simulator/src/basic_sim.rs:
```diff
@@ -16,7 +16,7 @@ use tokio::time::sleep;
 
 const END_EPOCH: u64 = 16;
-const GENESIS_DELAY: u64 = 32;
+const GENESIS_DELAY: u64 = 60;
```
We might be able to revert this now we use prod `c-kzg-4844` and `rust-eth-kzg` for PeerDAS, which only takes 500ms to initialize.

jimmygchen replied:
Done

jimmygchen commented on testing/ef_tests/check_all_files_accessed.py:
```diff
@@ -20,6 +20,10 @@
+    # TODO(das): remove once electra tests are on unstable
+    "tests/.*/electra/",
+    # TODO(das): ignore until new spec test release with column subnet count = 64.
+    "tests/.*/.*/.*/get_custody_columns/",
```
we should enable this after https://github.com/sigp/lighthouse/issues/6289

jimmygchen commented on testing/ef_tests/Makefile:
```diff
@@ -1,4 +1,4 @@
-TESTS_TAG := v1.5.0-alpha.2
+TESTS_TAG := v1.5.0-alpha.3
```
update to alpha.5

jimmygchen replied:
Will do this after merge as we may need to update some tests

jimmygchen commented on consensus/types/src/data_column_subnet_id.rs:
```diff
@@ -51,9 +51,9 @@ impl DataColumnSubnetId {
-                .expect("hash_fixed produces a 32 byte array");
+            let hash_prefix = [
+                hash[0], hash[1], hash[2], hash[3], hash[4], hash[5], hash[6], hash[7],
+            ];
```
I think this was accidentally reverted in a merge. We want the implementation on `unstable` (LHS).

jimmygchen replied:
Done

jimmygchen commented on beacon_node/src/cli.rs:
```diff
@@ -66,6 +66,25 @@ pub fn cli_app() -> Command {
                 .hide(true)
         )
+        .arg(
+            // TODO(das): remove this before release
```
```suggestion
            // TODO(das): remove this before PeerDAS release
```

michaelsproul committed:
Remove get_block_import_status

michaelsproul commented on beacon_node/beacon_chain/src/builder.rs:
```diff
@@ -409,6 +409,11 @@ where
                 .init_blob_info(genesis.beacon_block.slot())
                 .map_err(|e| format!("Failed to initialize genesis blob info: {:?}", e))?,
         );
+        self.pending_io_batch.push(
```
Could do this only if peer DAS is scheduled. Gives us a bit more freedom to change the on-disk format before the fork.

michaelsproul commented on beacon_node/beacon_chain/src/chain_config.rs:
```diff
@@ -84,6 +84,10 @@ pub struct ChainConfig {
     /// When set to true Light client server computes and caches state proofs for serving updates
     pub enable_light_client_server: bool,
+    /// Enable malicious PeerDAS mode where node withholds data columns when publishing a block
+    pub malicious_withhold_count: usize,
```
Could do with a more descriptive comment

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -360,64 +402,108 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
+            .filter_map(|block| block.custody_columns().cloned())
+            .flatten()
+            .map(CustodyDataColumn::into_inner)
+            .collect::<Vec<_>>();
```
we can probably add a condition here to avoid the allocation for the vec and list below if peerdas is not enabled

jimmygchen replied:
ah ignore this for now - this is a `Vec` of rpc blocks, so we've gotta check every block.

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -360,64 +402,108 @@ impl<T: BeaconChainTypes> DataAvailabilityChecker<T> {
+            fork_epoch,
+            current_slot
+                .epoch(T::EthSpec::slots_per_epoch())
+                .saturating_sub(self.spec.min_epochs_for_blob_sidecars_requests),
```
there's a separate config for data column sidecar requests

jimmygchen replied:
Let's address in https://github.com/sigp/lighthouse/issues/6240

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker.rs:
```diff
@@ -585,10 +662,15 @@ impl<E: EthSpec> AvailableBlock<E> {
         self.blobs.as_ref()
     }
 
+    #[allow(clippy::type_complexity)]
```
this can be removed

jimmygchen replied:
Done

jimmygchen commented on beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs:
```diff
@@ -18,16 +21,21 @@ use types::{
     Hash256, SignedBeaconBlock,
 };
 
+pub type DataColumnsToPublish<E> = Option<Vec<Arc<DataColumnSidecar<E>>>>;
```
```suggestion
pub type DataColumnsToPublish<E> = Option<DataColumnSidecarList>;
```

jimmygchen replied:
Done

michaelsproul commented on beacon_node/http_api/src/publish_blocks.rs:
```diff
@@ -104,13 +110,38 @@ pub async fn publish_block<T: BeaconChainTypes, B: IntoGossipVerifiedBlockConten
+                            .len()
+                            .saturating_sub(malicious_withhold_count);
+                        // Randomize columns before dropping the last malicious_withhold_count items
+                        data_col_sidecars.shuffle(&mut rand::thread_rng());
```
Could use a deterministic RNG here in testing (in future).

michaelsproul replied:
Opened an issue for later: https://github.com/sigp/lighthouse/issues/6296

michaelsproul commented on beacon_node/lighthouse_network/src/peer_manager/mod.rs:
```diff
@@ -892,6 +895,11 @@ impl<E: EthSpec> PeerManager<E> {
                 self.max_peers().saturating_sub(dialing_peers) - peer_count
             } else if outbound_only_peer_count < self.min_outbound_only_peers()
                 && peer_count < self.max_outbound_dialing_peers()
+                && self.target_peers > 10
+            // This condition is to attempt to exclude testnets without
+            // an explicit CLI flag. For networks with low peer counts, we don't want to do
+            // repetitive searches for outbound peers, when we may be already connected to every
+            // peer on the testnet
```
Pls delet

jimmygchen replied:
Done

michaelsproul commented on beacon_node/lighthouse_network/src/rpc/protocol.rs:
```diff
@@ -399,6 +399,9 @@ impl SupportedProtocol {
             supported.extend_from_slice(&[
                 ProtocolId::new(SupportedProtocol::BlobsByRootV1, Encoding::SSZSnappy),
                 ProtocolId::new(SupportedProtocol::BlobsByRangeV1, Encoding::SSZSnappy),
+                // TODO(das): add to PeerDAS fork
+                ProtocolId::new(SupportedProtocol::DataColumnsByRootV1, Encoding::SSZSnappy),
+                ProtocolId::new(SupportedProtocol::DataColumnsByRangeV1, Encoding::SSZSnappy),
```
Should delete this. It looks like it effects Deneb when it shouldn't.

jimmygchen replied:
Done

jimmygchen commented on beacon_node/lighthouse_network/src/service/mod.rs:
```diff
@@ -255,10 +257,11 @@ impl<E: EthSpec> Network<E> {
                 // during a fork we subscribe to both the old and new topics
                 max_subscribed_topics: max_topics * 4,
-                // 162 in theory = (64 attestation + 4 sync committee + 7 core topics + 6 blob topics) * 2
+                // 209 in theory = (64 attestation + 4 sync committee + 7 core topics + 6 blob topics + 64 column topics) * 2
```
```suggestion
                // 418 in theory = (64 attestation + 4 sync committee + 7 core topics + 6 blob topics + 128 column topics) * 2
```

jimmygchen commented on beacon_node/lighthouse_network/src/types/subnet.rs:
```diff
@@ -12,7 +12,7 @@ pub enum Subnet {
     /// Represents a gossipsub sync committee subnet and the metadata `syncnets` field.
     SyncCommittee(SyncSubnetId),
-    /// Represents a gossipsub data column subnet.
+    /// Represents a gossipsub data column subnet and the metadata `blbcolnets` field.
```
Remove the `blbcolnets` comment

jimmygchen replied:
Done

jimmygchen commented on beacon_node/network/src/sync/backfill_sync/mod.rs:
```diff
@@ -372,7 +372,9 @@ impl<T: BeaconChainTypes> BackFillSync<T> {
             // reasons. Check that this block belongs to the expected peer
-            if !batch.is_expecting_block(peer_id, &request_id) {
+            // TODO(das): removed peer_id matching as the node may request a different peer for data
+            // columns.
```
this may be a permanent change now - but i guess we can keep this until the range sync rework happens to be sure.

jimmygchen commented on beacon_node/network/src/sync/block_lookups/single_block_lookup.rs:
```diff
@@ -328,46 +329,45 @@ impl<E: EthSpec> BlobRequestState<E> {
             state: SingleLookupRequestState::new(),
         }
     }
 }
```
Looks like just re-ordering here, maybe revert this so we keep the git blames for those lines for the future.

jimmygchen commented on beacon_node/network/src/sync/block_lookups/single_block_lookup.rs:
```diff
@@ -328,46 +329,45 @@ impl<E: EthSpec> BlobRequestState<E> {
     /// Request is processing, sent by lookup sync
     Processing(DownloadResult<T>),
-    /// Request is processed
+    /// Request is processed:
```
:

jimmygchen commented on beacon_node/network/src/sync/block_sidecar_coupling.rs:
```diff
@@ -1,69 +1,105 @@
-use beacon_chain::block_verification_types::RpcBlock;
+use beacon_chain::{
```
file can probably be renamed to `range_block_components_request.rs`

jimmygchen replied:
I'll leave rename to later.

jimmygchen commented on beacon_node/network/src/sync/network_context.rs:
```diff
@@ -239,6 +285,22 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
     }
 
+    // TODO(das): epoch argument left here in case custody rotation is implemented
+    pub fn get_custodial_peers(&self, _epoch: Epoch, column_index: ColumnIndex) -> Vec<PeerId> {
```
`_epoch: Epoch` and the above coment can be removed

jimmygchen replied:
Done

jimmygchen commented on beacon_node/network/src/sync/network_context.rs:
```diff
@@ -695,6 +859,31 @@ impl<T: BeaconChainTypes> SyncNetworkContext<T> {
             });
     }
 
+    pub fn report_peer_on_rpc_error(&self, peer_id: &PeerId, error: &RPCError) {
+        // Note: logging the report event here with the full error display. The log inside
+        // `report_peer` only includes a smaller string, like "invalid_data"
+        debug!(self.log, "reporting peer for sync lookup error"; "error" => %error);
+        if let Some(action) = match error {
+            // Protocol errors are heavily penalized
+            RPCError::SSZDecodeError(..)
+            | RPCError::IoError(..)
+            | RPCError::ErrorResponse(..)
+            | RPCError::InvalidData(..)
+            | RPCError::HandlerRejected => Some(PeerAction::LowToleranceError),
+            // Timing / network errors are less penalized
+            // TODO: Is IoError a protocol error or network error?
+            RPCError::StreamTimeout | RPCError::IncompleteStream | RPCError::NegotiationTimeout => {
+                Some(PeerAction::MidToleranceError)
+            }
+            // Not supporting a specific protocol is tolerated. TODO: Are you sure?
+            RPCError::UnsupportedProtocol => None,
+            // Our fault, don't penalize peer
+            RPCError::InternalError(..) | RPCError::Disconnected => None,
+        } {
+            self.report_peer(*peer_id, action, error.into());
+        }
+    }
```
unused, delete

jimmygchen replied:
deleted

jimmygchen commented on beacon_node/network/src/sync/range_sync/chain.rs:
```diff
@@ -1087,6 +1107,36 @@ impl<T: BeaconChainTypes> SyncingChain<T> {
+    /// Checks all custody column subnets for peers. Returns `true` if there is at least one peer in
+    /// every custody column subnet.
+    fn good_peers_on_custody_subnets(&self, epoch: Epoch, network: &SyncNetworkContext<T>) -> bool {
+        if network.chain.spec.is_peer_das_enabled_for_epoch(epoch) {
```
We can probably lift the `if` condition up to the caller side, so it's not too misleading here

jimmygchen replied:
hmm realised  this is called in mutiple places, might be better to leave it here.

pawanjay176 replied:
Might be good to tweak the function docs to specify we check that peerdas is enabled

jimmygchen commented on beacon_node/network/src/sync/range_sync/chain.rs:
```diff
@@ -1117,6 +1167,15 @@ impl<T: BeaconChainTypes> SyncingChain<T> {
             return None;
         }
 
+        // don't send batch requests until we have peers on custody subnets
+        if !self.good_peers_on_custody_subnets(self.to_be_downloaded, network) {
+            debug!(
+                self.log,
+                "Waiting for peers to be available on custody column subnets"
+            );
+            return None;
+        }
```
This workaround can be removed once individual requests in range sync are made retryable in https://github.com/sigp/lighthouse/issues/6258, perhaps worth adding a comment here

jimmygchen replied:
Added comment

jimmygchen committed:
Merge branch 'unstable' into das

jimmygchen committed:
Re-enable Windows release tests.

jimmygchen committed:
Address some review comments.

jimmygchen committed:
Address more review comments and cleanups.

michaelsproul committed:
Comment out peer DAS KZG EF tests for now

jimmygchen reviewed (approved):
Looks good to me! Let's go🚀

jimmygchen committed:
Address more review comments and fix build.

jimmygchen committed:
Merge branch 'das' of github.com:sigp/lighthouse into das

jimmygchen commented:
btw if we're merging this PR with mergify, mergify will delete the `das` branch and automatically  close all PRs targeting `das`:
https://github.com/sigp/lighthouse/pulls?q=is%3Apr+is%3Aopen+label%3Adas

realbigsean commented on testing/ef_tests/check_all_files_accessed.py:
```diff
@@ -20,6 +20,10 @@
 # *not* running the spec tests).
 excluded_paths = [
+    # TODO(das): remove once electra tests are on unstable
+    "tests/.*/electra/",
```
I think we should make this more targeted, it could make us miss other electra tests

michaelsproul replied:
they actually all run fine, so deleted in 45b3203e5d94a54465ad80e8f14b7b7ed87db9a0

realbigsean commented on beacon_node/beacon_chain/src/blob_verification.rs:
```diff
@@ -409,8 +409,8 @@ pub fn validate_blob_sidecar_for_gossip<T: BeaconChainTypes>(
         return Err(GossipBlobError::InvalidSubnet {
-            expected: blob_index,
-            received: subnet,
+            expected: subnet,
```
good catch!

realbigsean commented on beacon_node/beacon_processor/src/metrics.rs:
```diff
@@ -221,6 +245,15 @@ pub static BEACON_PROCESSOR_REPROCESSING_QUEUE_MATCHED_ATTESTATIONS: LazyLock<Re
+    Result<IntCounter>,
+> = LazyLock::new(|| {
+    try_create_int_counter(
+        "beacon_processor_reprocessing_queue_matches_sampling_requests",
```
The variable name doesn't match the metric name, not sure which was intended
```suggestion
        "beacon_processor_reprocessing_queue_matched_sampling_requests",
```

realbigsean commented on beacon_node/lighthouse_network/src/service/api_types.rs:
```diff
@@ -42,11 +42,43 @@ pub enum SyncRequestId {
+#[derive(Debug, Hash, PartialEq, Eq, Clone, Copy)]
+pub struct CustodyId {
+    pub requester: CustodyRequester,
+    pub req_id: Id,
```
should this be a `CustodyRequestId` newtype similar to `SamplingRequestId`?

realbigsean commented on beacon_node/lighthouse_network/src/types/globals.rs:
```diff
@@ -111,14 +119,45 @@ impl<E: EthSpec> NetworkGlobals<E> {
+
+    /// Returns a connected peer that:
+    /// 1. is connected
+    /// 2. assigned to custody the column based on it's `custody_subnet_count` from ENR or metadata (WIP)
```
is this still a WIP?

dapplion replied:
Metadatav3 completes this wip

realbigsean commented on beacon_node/network/src/sync/sampling.rs:
```diff
@@ -0,0 +1,628 @@
+    }
+}
+
+mod request {
```
separate file?

realbigsean commented:
Had some relatively inconsequential feed back but looks good overall! 🚀

pawanjay176 commented on scripts/local_testnet/network_params_das_devnet_1.yaml:
Not sure if the devnet config file should go to unstable

michaelsproul replied:
agree, @jimmygchen can we delete this?

pawanjay176 commented on consensus/types/src/runtime_var_list.rs:
```diff
@@ -1,20 +1,58 @@
-impl<T: Encode + Decode + Clone> RuntimeVariableList<T> {
-    pub fn new(vec: Vec<T>, max_len: usize) -> Result<Self, ssz_types::Error> {
+impl<T> RuntimeVariableList<T> {
+    /// Returns `Some` if the given `vec` equals the fixed length of `Self`. Otherwise returns
```
Some -> Ok
None -> OutOfBounds error

michaelsproul replied:
updated in 77caa971310387e1b6ff106aaf3e168e30c0c726

pawanjay176 commented on beacon_node/src/cli.rs:
```diff
@@ -66,6 +66,25 @@ pub fn cli_app() -> Command {
                 .hide(true)
         )
+        .arg(
+            // TODO(das): remove this before PeerDAS release
```
I assume this is before the user facing release with peerdas fork params set in, but just confirming that we don't need to remove this now?

michaelsproul replied:
I think it's fine to stay, as it's a hidden flag

dapplion replied:
There's no network a user can attack as of now

michaelsproul committed:
Unignore Electra tests

pawanjay176 commented on beacon_node/beacon_chain/src/metrics.rs:
```diff
@@ -1619,6 +1619,13 @@ pub static BLOBS_SIDECAR_PROCESSING_REQUESTS: LazyLock<Result<IntCounter>> = Laz
         "Count of all blob sidecars submitted for processing",
     )
 });
+pub static BLOBS_COLUMN_SIDECAR_PROCESSING_REQUESTS: LazyLock<Result<IntCounter>> =
```
Can this be deleted? Its not used anywhere

michaelsproul replied:
done

michaelsproul committed:
Fix metric name

pawanjay176 commented on beacon_node/lighthouse_network/src/peer_manager/peerdb.rs:
```diff
@@ -746,8 +794,17 @@ impl<E: EthSpec> PeerDB<E> {
+                        &self.spec,
+                    )
+                    .collect::<HashSet<_>>();
+                    info.set_custody_subnets(custody_subnets);
```
just noting that we'll have a list of custody subnets for all our peers on mainnet too with this.
Don't see anything harmful with this though other than some extra memory

pawanjay176 reviewed (changes requested):
Nothing major, gonna go over for a few more files tomorrow

michaelsproul committed:
Address some of Pawan's review comments

dapplion commented on beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs:
```diff
@@ -430,6 +449,28 @@ impl<T: BeaconChainTypes> DataAvailabilityCheckerInner<T> {
         }
     }
 
+    /// Potentially trigger reconstruction if:
+    /// - Our custody requirement is all columns
+    /// - We >= 50% of columns, but not all columns
```
Source code below is self-explanatory enough
```suggestion
    /// Returns true if reconstruction should start. Note that reconstruction is started only once.
    /// Currently only supernodes expecting all columns will attempt reconstruction.
```

dapplion commented on beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs:
```diff
@@ -486,19 +530,68 @@ impl<T: BeaconChainTypes> DataAvailabilityCheckerInner<T> {
 
         let block_import_requirement = self.block_import_requirement(epoch)?;
+
+        // Potentially trigger reconstruction if:
+        // - Our custody requirement is all columns
+        // - We >= 50% of columns
```
```suggestion
```

dapplion commented on beacon_node/beacon_chain/src/metrics.rs:
```diff
@@ -1653,6 +1653,13 @@ pub static DATA_COLUMN_SIDECAR_COMPUTATION: LazyLock<Result<Histogram>> = LazyLo
+    LazyLock::new(|| {
+        try_create_histogram(
+            "data_column_sidecar_inclusion_proof_verification_seconds",
+            "Time taken to verify data_column sidecar inclusion proof",
```
Should add proper bucket values to the new histograms

dapplion commented on beacon_node/beacon_processor/src/work_reprocessing_queue.rs:
```diff
@@ -133,6 +143,12 @@ pub struct QueuedLightClientUpdate {
+pub struct QueuedSamplingRequest {
+    pub beacon_block_root: Hash256,
+    pub process_fn: BlockingFn,
+}
```
Need more metrics around this mechanism to compute
- ratio of dropped requests
- time of arrival into the slot
- time to fulfill into the slot

pawanjay176 commented on beacon_node/beacon_chain/src/data_availability_checker/overflow_lru_cache.rs:
```diff
@@ -486,19 +530,68 @@ impl<T: BeaconChainTypes> DataAvailabilityCheckerInner<T> {
+                // Will only return an error if:
+                // - < 50% of columns
+                // - There are duplicates
+                let all_data_columns = KzgVerifiedCustodyDataColumn::reconstruct_columns(
```
Since this is a potentially long running blocking task, should we run it on a Blocking function instead of async in `Work::GossipDataColumnSidecar`?

jimmygchen replied:
Yeah good call - I've got a PR that refactors the reconstruction code, and extracts the logic out of DA checker, it might be easier to include the change there I think
https://github.com/sigp/lighthouse/pull/5990

pawanjay176 replied:
Sounds good 👍

pawanjay176 reviewed (approved):
LGTM as a whole. My main focus was testing that merging this doesn't affect mainnet nodes adversely. 
Will review sampling and custody logic on peerdas devnets.
Tested range sync and lookup sync on mainnet in different scenarios and haven't faced any issues. Great work on pushing this through 🙌 
Just one question.

michaelsproul committed:
Merge remote-tracking branch 'origin/unstable' into das

eserilev committed:
Update PeerDAS network parameters for peerdas-devnet-2  (#6290)

* update subnet count & custody req

* das network params

* update ef tests

---------

Co-authored-by: Jimmy Chen <jchen.tc@gmail.com>

michaelsproul commented:
@mergify queue

dapplion reviewed (commented):
So happy to see this merged 🎉 big thanks to the reviewers and @jimmygchen 

Some post-merge comments
